### ISSUE TITLE
Rule 901410 uses @rx, when an @unconditionalMatch would work the same way

### ISSUE THREAD

**Issue by dune73** [coreruleset/coreruleset#2091]
Simple first issue saving a bit of performance (when using the Sampling Mode that is mainly about saving performance).

**Comment by 53cur3M3** [comment]
Would you like me to submit a pull request for this, or leave this as a first issue for someone else?

**Comment by azurit** [comment]
@53cur3M3 It would be cool if you'll send a PR, thank you!

**Comment by 53cur3M3** [comment]
I intially made this simple change... but then wondered if the sampling mode rules could be made more efficient still, the likelihood of not getting two integers out of a hex-encoded sha1 string is statistically highly improbable, so I came up with the following which reduces the number of regexes. I'm assuming that a length comparison is more efficient than a regex match:

`SecRule TX:sampling_percentage "@eq 100" \
    "id:901400,\
    phase:1,\
    pass,\
    nolog,\
    ver:'OWASP_CRS/3.3.0',\
    skipAfter:END-SAMPLING"
SecRule UNIQUE_ID "@rx ^[a-f]*(0|([1-9]))[a-f]*([0-9])" \
    "id:901410,\
    phase:1,\
    pass,\
    t:sha1,t:hexEncode,\
    capture,\
    nolog,\
    ver:'OWASP_CRS/3.3.0',\
    setvar:'TX.sampling_rnd100=%{TX.2}%{TX.3}'"
SecRule TX:sampling_rnd100 "!@gt 0" \
    "id:901420,\
    phase:1,\
    pass,\
    t:length,\
    log,\
    msg:'{MATCHED_VAR} Using microseconds from duration as did not get integers from hexEncoded hash',\
    ver:'OWASP_CRS/3.3.0',\
    chain"
        SecRule DURATION "@rx ([1-9]?[0-9])$" \
        "capture,\
        setvar:'TX.sampling_rnd100=%{TX.1}'"`

I functionally tested (but haven't performance tested) the above ( https://github.com/53cur3M3/coreruleset/tree/improveSamplingEfficiency ) in apache modsec2 and it seemed to behave as expected.

I sanity checked this with @dune73 to make sure there wasn't a really glaring problem. He then took it a step further:
"_The interesting question is what happens to 901450 when TX:sampling_rnd100 is empty. And I just checked, the rule does not trigger. So in the rare case that we do not get digits out of the hash, we simply execute the rule set which I think is the desired behaviour.
I also checked what happens if we keep the leading 0 in the variable and 901450 executes just fine. So it looks like this could be simplified substantially._"

In which case, I think we could potentially reduce this to:
`SecRule TX:sampling_percentage "@eq 100" \
    "id:901400,\
    phase:1,\
    pass,\
    nolog,\
    ver:'OWASP_CRS/3.3.0',\
    skipAfter:END-SAMPLING"
SecRule UNIQUE_ID "@rx ^[a-f]*([0-9])[a-f]*([0-9])" \
    "id:901410,\
    phase:1,\
    pass,\
    t:sha1,t:hexEncode,\
    capture,\
    nolog,\
    ver:'OWASP_CRS/3.3.0',\
    setvar:'TX.sampling_rnd100=%{TX.1}%{TX.2}'"`

Would be interested in other peoples thoughts, and whether anyone would be willing to test this against modsec3-nginx to check that if it behaves the same?

**Comment by 53cur3M3** [comment]
sorry for above code copy & paste fail - hasn't rendered newlines

**Comment by azurit** [comment]
Don't worry, i fixed it.

**Comment by azurit** [comment]
If i understand it correctly, you are replacing rules 901410, 901420, 901430 and 901440 with your 901410, are you?

**Comment by 53cur3M3** [comment]
@azurit , yes that's correct (or replacement of 901410-901440 with revised 901410 and 901420 - if we did want to keep fallback to duration).

**Comment by dune73** [comment]
When I did this code year's ago I wanted to make it perfect. I traded simplicity and performance for perfection, while admitting in the comments that the randomness is far from perfect (but good enough for sampling).

With more distance, I think @53cur3M3 is right in favoring simplicity and performance over perfection (that can not be achieved here).

Just to add a bit perspective: The whole duration hack comes into play, when a hexEncoded SHA1 hash consists of a-f alone, thus does not contain any digits. The chance of this happening is 0.375 ** 20 if I am not mistaken. That's a change of 0.000000003024 and I think that is pretty rare. So might just as well skip it.
